server-spi-private/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantType.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.protocol.oidc.grants;
+
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.MultivaluedMap;
+import jakarta.ws.rs.core.Response;
+
+import java.util.Map;
+
+import org.keycloak.common.ClientConnection;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.http.HttpRequest;
+import org.keycloak.http.HttpResponse;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.provider.Provider;
+import org.keycloak.provider.ProviderFactory;
+import org.keycloak.representations.dpop.DPoP;
+
+/**
+ * Provider interface for OAuth 2.0 grant types
+ *
+ * @author <a href="mailto:demetrio@carretti.pro">Dmitry Telegin</a>
+ */
+public interface OAuth2GrantType extends Provider, ProviderFactory<OAuth2GrantType> {
+
+    /**
+     * Returns EventType associated with this grant type
+     *
+     * @return event type
+     */
+    EventType getEventType();
+
+    /**
+     * Returns the name of the OAuth 2.0 grant type implemented by this provider.
+     * This value will be matched against the "grant_type" token request parameter.
+     *
+     * @return grant type name
+     */
+    String getGrantType();
```

Comment:
What if we also support this https://github.com/keycloak/keycloak/pull/24512 from this SPI?